Implement more qlog traces: CC and Recovery state

agrover opened the pull request:
More interesting values, such as from congestion control and recovery.

The early commits in this PR stick with allowing qlog tracing functions to return `Res`, but this is converted later to handling qlog errors internally (response: emit a logmessage and close the log) rather than imposing error-handling on calling functions.

One burden calling functions *do* have to bear is ensuring the `NeqoQlog` is in scope. It's refcounted so that helps, but it does require a couple additional `set_qlog` methods.

agrover commented on neqo-transport/src/cc.rs:
```diff
@@ -87,22 +103,32 @@ impl CongestionControl {
             }
             if self.app_limited() {
                 // Do not increase congestion_window if application limited.
+                qlog::congestion_state_updated(&mut self.qlog.borrow_mut(), "application_limited");
```
good idea.

agrover commented on neqo-transport/src/qlog.rs:
```diff
@@ -166,9 +226,103 @@ fn connection_started(qlog: &mut Option<NeqoQlog>, path: &Path) -> Res<()> {
+            }
+        }
+
+        let res = qlog.stream().add_event(Event::metrics_updated(
+            min_rtt,
+            smoothed_rtt,
+            latest_rtt,
+            rtt_variance,
+            max_ack_delay,
+            pto_count,
+            congestion_window,
+            bytes_in_flight,
+            ssthresh,
+            packets_in_flight,
+            in_recovery,
+            pacing_rate,
+        ));
```
lol, I'll see what they say.

agrover commented on neqo-transport/src/qlog.rs:
```diff
@@ -79,80 +85,134 @@ pub fn connection_tparams_set(
-) -> Res<()> {
-    if let Some(qlog) = qlog {
+) {
+    if let Some(qlog) = maybe_qlog {
```
I'm ok with a little complexity in qlog.rs files if it keeps complexity out of qlog callsites in the "real" code, and also if it keeps work done if qlog not enabled to ~zero. I don't see a way to do what you're describing without hurting those two other criteria, but I'd be happy to be proven wrong.

martinthomson replied:
See above.  The repetition here is avoidable.

agrover commented on neqo-transport/src/qlog.rs:
```diff
@@ -79,80 +85,134 @@ pub fn connection_tparams_set(
+            ))?;
+
+            while d.remaining() > 0 {
+                match Frame::decode(&mut d) {
```
This only occurs if qlog is enabled. This also keeps the probe count in the real code to a minimum (one).

agrover commented:
Rebased, and added commits:
* qlog: Implement h3 data moved up/down
* Change NeqoQlog::new() to take a ref instead of an owned PathBuf
* Changes based on initial review

agrover commented on neqo-transport/src/crypto.rs:
```diff
@@ -1030,6 +1030,6 @@ impl Default for CryptoStreams {
     space: PNSpace,
-    offset: u64,
-    length: usize,
+    pub offset: u64,
+    pub length: usize,
```
in qlog::packet_lost we need this to call `QuicFrame::crypto()`. I thought since CryptoRecoveryToken was pretty much a group of values that just making them pub was reasonable, but if you want accessor fns we can do that too.

martinthomson replied:
Now you made me look closer at that, I think that it is very badly wrong to be doing this.  See the comment about that above.

agrover commented on neqo-transport/src/qlog.rs:
```diff
@@ -166,9 +226,103 @@ fn connection_started(qlog: &mut Option<NeqoQlog>, path: &Path) -> Res<()> {
+    }
+}
+
+thread_local!(static CURR_CONG_STATE: RefCell<Option<String>> = RefCell::new(None));
```
I have no idea what substance I was on when I wrote this. Yes this needs to be per-connection.

martinthomson commented on neqo-transport/src/cc.rs:
```diff
@@ -87,22 +105,44 @@ impl CongestionControl {
             self.bytes_in_flight -= pkt.size;
 
-            if self.in_congestion_recovery(pkt.time_sent) {
+            if self.in_congestion_recovery(pkt.time_sent, PacketState::Acked) {
```
OK, this is a functionality change.  This clears the recovery timer.  I think that's right, but I'm concerned that I can't find a test that has updated as a result of the change.  That means that we weren't testing the exit from congestion recovery properly.

martinthomson commented on neqo-transport/src/connection.rs:
```diff
@@ -2084,6 +2093,7 @@ impl Connection {
     /// to retransmit the frame as needed.
     fn handle_lost_packets(&mut self, lost_packets: &[SentPacket]) {
         for lost in lost_packets {
+            qlog::packet_lost(&mut self.qlog.borrow_mut(), &lost);
```
Unfortunately, this is going to be wrong.  We use this function for marking data as lost for reasons other than lost packets.  There are two places that we genuinely mark packets as lost: when we handle an ACK and when we run the time-based loss recovery.  This is also called for discarded 0-RTT, PTO, and Retry.

However, I also think that qlog is completely wrong in what it is asking for here.  It should be recording just the packet number that was marked lost and relying on the logging for that packet to determine what data was lost along with that.

agrover replied:
ok, will do.

agrover commented on neqo-http3/src/qlog.rs:
```diff
@@ -0,0 +1,39 @@
+}
+
+pub fn h3_data_moved_down(maybe_qlog: &mut Option<NeqoQlog>, stream_id: u64, amount: usize) {
+    if let Some(qlog) = maybe_qlog {
+        let res = qlog.stream().add_event(Event::h3_data_moved(
+            stream_id.to_string(),
+            None,
+            Some(u64::try_from(amount).unwrap()),
+            Some(H3DataRecipient::Application),
+            Some(H3DataRecipient::Transport),
+            None,
+        ));
+        handle_qlog_result(maybe_qlog, res)
+    }
```
Yeah let's do this separately. 

I think I get what you're saying but my thought was it is better to check qlog is enabled before generating the event. Doesn't your suggestion above flip this so the Event is constructed to be passed to `add_event` but then discarded if !qlog? Some of these events are rather big and I'm hesitant to impose the cost of constructing when the far more common case is that qlog isn't enabled.

agrover commented:
Ok, updated/rebased:
* Exiting congestion recovery is covered by `cc_cong_avoidance_recovery_period_to_cong_avoidance` test.
* Changed `packet_lost` to not log frames, and to not log "lost" frames that are 0rtt, PTO, retry. (No longer making undesirable `pub` changes to token structs)
* No changes for wrapper, we agreed to defer this.
